docs/source/deprecation.md

```diff
@@ -1,17 +1,458 @@
+for the test function, we will get error (see that we have now just one
+deprecation warning now):
+
+```python
+| 21:59:57 | conftest | INFO | conftest.py | Actor 'foo_producer' context teardown complete
+
+repos/system_upgrade/el7toel8/actors/fooproducer/tests/test_unit_fooproducer.py::test_process FAILED
+
+====================================================== FAILURES ======================================================
+____________________________________________________ test_process ____________________________________________________
+
+args = (), kwargs = {'monkeypatch': <_pytest.monkeypatch.MonkeyPatch object at 0x7f21924b24d0>}
+suppressed = <class 'leapp.models.foo.Foo'>
+
+    @functools.wraps(target_item)
+    def process_wrapper(*args, **kwargs):
+        for suppressed in suppressed_items:
+            _suppressed_deprecations.add(suppressed)
+        try:
+            return target_item(*args, **kwargs)
+        finally:
+            for suppressed in suppressed_items:
+>               _suppressed_deprecations.remove(suppressed)
+E               KeyError: <class 'leapp.models.foo.Foo'>
+
+tut/lib/python3.7/site-packages/leapp/utils/deprecation.py:35: KeyError
+================================================== warnings summary ==================================================
+repos/system_upgrade/el7toel8/actors/fooproducer/tests/test_unit_fooproducer.py::test_process
+  /tmp/leapp-repository/repos/system_upgrade/el7toel8/actors/fooproducer/tests/test_unit_fooproducer.py:13: _DeprecationWarningContext: Usage of deprecated Model "Foo"
+    assert Foo(value='Answer is: 42') in produced_msgs.model_instances
+
+-- Docs: http://doc.pytest.org/en/latest/warnings.html
+======================================== 1 failed, 1 warnings in 0.21 seconds ========================================
+```
+
+It's because the mechanism of decorators in python and how pytest works.
+In this case, we need to do a small workaround, like this:
+
+```python
+from leapp.libraries.actor import fooproducer_lib
+from leapp.libraries.common.testutils import produce_mocked
+from leapp.utils.deprecation import suppress_deprecation
+from leapp.libraries.stdlib import api
+from leapp.models import Foo  # deprecated model
+
+
+@suppress_deprecation(Foo)
+def _foo(value):
+    """Small workaround to suppress deprecation messages in tests."""
+    return Foo(value=value)
+
+def test_process(monkeypatch):
+    produced_msgs = produce_mocked()
+    monkeypatch.setattr(api, 'produce', produced_msgs)
+    fooproducer_lib.produce_answer()
+    assert _foo(value='Answer is: 42') in produced_msgs.model_instances
+```
+
+That's the whole solution for the *FooProducer* actor. Analogically to this,
+we need to treat the `FooConsumer` actor. You could notice, that all imports
+of the `Foo` model are commented. It's good practice as it is more visible
+to all developers, that a deprecated entity is present.
+
+### Example of a derived model deprecation
```

Comment:
All this chapter is very confusing to me. I think this is absolutely ok to deprecate the base model and expect that children will fire the deprecation warning. The only thing that we can do here is to improve the message of the deprecation warning so that it will contain the real source of the deprecation. but this is just a feature improvement to the deprecation functionality.

Reply:
Basically the thing you suggest is not possible without actions above the parser of the python code. That's definitely not something we want to do as we have no time right now. We have been thinking about the solution and really do not have a good one so far. This is the best we have and people have to live with that until we have another.

Btw, what is confusing on this section? How it can be improved?

Reply:
Skipping from commenting:
> We have been thinking about the solution and really do not have a good one so far

But something like https://gist.github.com/ZhukovGreen/10544d9ed7150f450056b1791616c626 will work

Reply:
As a simple hack, we might just update the deprecation message to look something like this:
before
```
Usage of deprecated Model "BaseFoo"
```
after
```
Usage of deprecated Model "BaseFoo" or models inherited from it
```

Reply:
> But something like https://gist.github.com/ZhukovGreen/10544d9ed7150f450056b1791616c626 will work

Currently, we need to write documentation regarding the functionality in the released version. The proposal seems nice, but it's not part of the release, so it could be improvement for the future. As well, we have to have versioned documentation vefore we start to do updates of doc regarding the upstream devel changes.

Reply:
ok